airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/streams/async_job.py

```diff
@@ -0,0 +1,277 @@
+        """Start each job in the group."""
+        for job in self._jobs:
+            job.start()
+
+    def restart(self):
+        """Restart failed jobs"""
+        for job in self._jobs:
+            if job.failed:
+                job.restart()
+            self._restart_number = max(self._restart_number, job.restart_number)
+
+    @property
+    def restart_number(self):
+        """Number of restarts"""
+        return self._restart_number
+
+    @property
+    def completed(self) -> bool:
+        """Check job status and return True if all jobs are completed, use failed/succeeded to check if it was successful"""
+        return all(job.completed for job in self._jobs)
+
+    @property
+    def failed(self) -> bool:
+        """Tell if any job previously failed"""
+        return any(job.failed for job in self._jobs)
+
+    def update_job(self, batch=None):
+        """Checks jobs status in advance and restart if some failed."""
+        batch = self._api.new_batch()
+        unfinished_jobs = [job for job in self._jobs if not job.completed]
+        for jobs in chunks(unfinished_jobs, 50):
+            for job in jobs:
+                job.update_job(batch=batch)
+
+            while batch:
+                # If some of the calls from batch have failed, it returns  a new
+                # FacebookAdsApiBatch object with those calls
+                batch = batch.execute()
+
+    def get_result(self) -> Any:
+        """Retrieve result of the finished job."""
+        for job in self._jobs:
+            yield from job.get_result()
+
+    def split_job(self) -> "AsyncJob":
+        """Split existing job in few smaller ones grouped by ParentAsyncJob class"""
+        raise RuntimeError("Splitting of ParentAsyncJob is not allowed.")
+
+
+class InsightAsyncJob(AsyncJob):
+    """AsyncJob wraps FB AdReport class and provides interface to restart/retry the async job"""
+
+    def __init__(self, api, edge_object: Any, params: Mapping[str, Any], key: Optional[Any] = None):
+        """Initialize
+
+        :param api: FB API
+        :param edge_object: Account, Campaign, (AdSet or Ad in future)
+        :param params: job params, required to start/restart job
+        """
+        self._api = api
+        self._params = params
+        self._edge_object = edge_object
+        self._job: Optional[AdReportRun] = None
+        self._start_time = None
+        self._finish_time = None
+        self._failed = False
+        self._restart_number = 0
+        self.key = key
+
+    def split_job(self) -> ParentAsyncJob:
+        """Split existing job in few smaller ones grouped by ParentAsyncJob class.
+        TODO: use some cache to avoid expensive queries across different streams.
+        """
+        campaign_params = dict(copy.deepcopy(self._params))
+        # get campaigns from attribution window as well (28 day + 1 current day)
+        new_start = pendulum.parse(self._params["time_range"]["since"]) - pendulum.duration(days=28 + 1)
```

Comment:
is it always guaranteed that time_range and timerange.since will be set? string literal access makes me a bit nervous here

Reply:
it is, but let me change this